base/reducedim.jl

```diff
@@ -27,271 +37,323 @@ function reduced_indices0(axs::Indices{N}, region) where N
+    _mapreducedim_impl(x->F(x), op, init, A::AbstractArrayOrBroadcasted, raxs, out)
+function _mapreducedim_impl(f, op, init, A::AbstractArrayOrBroadcasted, raxs, out)
+    # @show f, op, init, A, raxs, out
+    outer_inds = CartesianIndices(raxs)
+    a_inds = CartesianIndices(axes(A))
+    Aaxs = axes(A)
+    @assert length(Aaxs) == length(raxs)
+    # The "outer dimension" of the reduction are the preserved dimensions
+    is_outer_dim = map((>(1))∘length, raxs)
+    n = prod(map((b, ax)->ifelse(b, 1, length(ax)), is_outer_dim, Aaxs))
+
+    # We have special support for returning empty arrays
+    # isempty(outer_inds) && return collect_allocator(_mapreduce_similar_curried(A),
+    #     (_mapreduce_naive_inner_loop(f, op, init, A, is_outer_dim, outer_ind, a_inds) for outer_ind in outer_inds))
+    (n == 0 || isempty(A)) && return _mapreduce_empty_array(f, op, init, A, raxs, out)
+    n == 1 && return _mapreduce_one_array(f, op, init, A, raxs, out)
+
+    _mapreduce_might_widen(f, op, A, out) && return collect_allocator(_mapreduce_similar_curried(A),
+        (_mapreduce_naive_inner_loop(f, op, init, A, is_outer_dim, outer_ind, a_inds) for outer_ind in outer_inds))
+
+    # Create the result vector with the first step of the reduction, using the passed output if given
+    # note that this is using a (potentially) bad cache ordering, so we don't want to continue like this
+    i1 = first(outer_inds)
+    inner1 = mergeindices(is_outer_dim, i1, a_inds)
+    i1, i2 = inner1
+    @inbounds a1, a2 = A[i1], A[i2]
+    v = _mapreduce_start(f, op, init, a1, a2)
+    R = isnothing(out) ? mapreduce_similar(A, typeof(v), raxs) : out
+    @inbounds R[i1] = v
+    for i in Iterators.drop(outer_inds, 1)
+        innerj = mergeindices(is_outer_dim, i, a_inds)
+        j1, j2 = innerj
+        @inbounds c1, c2 = A[j1], A[j2]
+        @inbounds R[i] = _mapreduce_start(f, op, init, c1, c2)
+    end
+    # If we only had two elements we can quickly return
+    # n == 2 && return R
+
+    lsiz = _linear_reduction_length(A, raxs)
+    if has_fast_linear_indexing(A) && lsiz > 0
+        ibase = first(LinearIndices(A))-1
+        if lsiz > 16
+            # use mapreduce_impl, which is better tuned to achieve higher performance and works pairwise
+            for i in eachindex(R)
+                @inbounds R[i] = op(R[i], mapreduce_impl(f, op, A, ibase+3, ibase+lsiz))
+                ibase += lsiz
+            end
+        else
+            for i in eachindex(R)
+                @inbounds r = R[i]
+                @simd for Ai in ibase+3:ibase+lsiz
+                    r = op(r, f(@inbounds(A[Ai])))
+                end
+                @inbounds R[i] = r
+                ibase += lsiz
+            end
+        end
+    else
+        # Take care of the smallest cartesian "rectangle" that includes our two peeled elements
+        small_inner, large_inner = _split2(inner1)
```

Comment:
This optimization is not valid if `_split2` decides not to split the last reduced dimensions, as
```julia
julia> I1 = collect(Iterators.flatten([CartesianIndices((1:2, 10)), CartesianIndices((3:10, 10))]));

julia> I2 = collect(Iterators.flatten([CartesianIndices((10, 1:1)), CartesianIndices((10, 2:10))]));

julia> I1 == I2
false

julia> sort(I1) == I2
true
```
A better solution might be skipping the first 2 element manually, e.g.
```julia
        indsAt, indsRt = safe_tail(Aaxs), safe_tail(raxs)
        keep, Idefault = Broadcast.shapeindexer(indsRt)
        skip = mergeindices(is_outer_dim, a_inds, i1:i2)
        skipt = Base.IteratorsMD.split(skip, Val(1))[2]
        if length(raxs[1]) == 1 && length(Aaxs[1]) > 1 #reducedim1(R, A)
            i1 = first(raxs[1])
            @inbounds for IA in CartesianIndices(indsAt)
                IR = Broadcast.newindex(IA, keep, Idefault)
                r = R[i1,IR]
                offset = IA in skipt ? 2 : 0
                @simd for i in Aaxs[1][begin + offset:end]
                    r = op(r, f(A[i, IA]))
                end
                R[i1,IR] = r
            end
        else
            @inbounds for IA in CartesianIndices(indsAt)
                IA in skipt && continue
                IR = Broadcast.newindex(IA, keep, Idefault)
                @simd for i in Aaxs[1]
                    R[i,IR] = op(R[i,IR], f(A[i,IA]))
                end
            end
        end
```